File: app/src/main/java/uk/co/transferx/app/ui/signin/fragment/SignInPinFragment.java
package uk.co.transferx.app.ui.signin.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.chaos.view.PinView;

import javax.inject.Inject;

import uk.co.transferx.app.ui.base.BaseFragment;
import uk.co.transferx.app.R;
import uk.co.transferx.app.TransferXApplication;
import uk.co.transferx.app.ui.mainscreen.MainActivity;
import uk.co.transferx.app.ui.signin.presenter.SignInPinPresenter;
import uk.co.transferx.app.ui.welcom.WelcomeActivity;

/**
 * Created by sergey on 19.11.17.
 */

public class SignInPinFragment extends BaseFragment implements SignInPinPresenter.SignInPinUI {

    private PinView pinView;
    private TextView label;
    private TextWatcher textWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            setStatusOfError(pinView, label, R.color.black);
            setLinesColor(R.color.black);
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    };

    @Override
    public String tagName() {
        return SignInPinFragment.class.getSimpleName();
    }

    @Inject
    SignInPinPresenter presenter;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        ((TransferXApplication) getActivity().getApplication()).getAppComponent().inject(this);
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.signin_pin_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        pinView = view.findViewById(R.id.first_pin);
        pinView.setAnimationEnable(true);
        ((TextView) view.findViewById(R.id.title)).setText(R.string.enter_your_pin);
        label = view.findViewById(R.id.first_label_pin);
        view.findViewById(R.id.sign_in).setOnClickListener(v -> presenter.checkPingAndLogIn(pinView.getText().toString()));
        view.findViewById(R.id.forgot_pin).setOnClickListener(v -> presenter.resetPassword());
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.attachUI(this);
        pinView.addTextChangedListener(textWatcher);
    }

    @Override
    public void onPause() {
        presenter.detachUI();
        super.onPause();
        pinView.removeTextChangedListener(textWatcher);
    }

    @Override
    public void goToMainScree() {
        Activity activity = getActivity();
        if (activity != null) {
            MainActivity.startMainActivity(activity);
            activity.finish();
        }
    }

    @Override
    public void pinChecked() {
        hideKeyboard(pinView);
        getActivity().finish();
    }

    private void setLinesColor(@ColorRes int color) {
        pinView.setLineColor(ContextCompat.getColor(pinView.getContext(), color));
    }

    @Override
    public void showError(String message) {
        Toast.makeText(getContext(), "Somthing went wrong ", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void goToWelcomeScreen() {
        hideKeyboard(pinView);
        WelcomeActivity.startWelcomeActivity(getActivity());
    }

    @Override
    public void showErrorPin() {
        setStatusOfError(pinView, label, R.color.red);
        setLinesColor(R.color.red);
    }

    @Override
    public void goToWelcome() {
        //no op
    }
}

File: app/src/main/java/uk/co/transferx/app/ui/signinpin/fragment/SignInPinFragment.kt
package uk.co.transferx.app.ui.signinpin.fragment

import android.graphics.Color
import android.os.Bundle
import android.support.annotation.ColorRes
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.chaos.view.PinView
import kotlinx.android.synthetic.main.pin_content.*
import kotlinx.android.synthetic.main.signin_fragment_layout.*
import kotlinx.android.synthetic.main.signin_pin_fragment.*
import org.jetbrains.anko.clearTask
import org.jetbrains.anko.intentFor
import org.jetbrains.anko.newTask
import uk.co.transferx.app.R
import uk.co.transferx.app.TransferXApplication
import uk.co.transferx.app.ui.base.BaseFragment
import uk.co.transferx.app.ui.homescreen.MainActivity
import uk.co.transferx.app.ui.signinpin.presenter.SignInPinPresenter
import uk.co.transferx.app.util.Constants
import uk.co.transferx.app.util.Constants.CLEAN_START
import javax.inject.Inject

/**
 * Created by sergey on 19.11.17.
 * Refactored and redesigned by Catalin Ghita on 11/2018
 */

class SignInPinFragment : BaseFragment(), SignInPinPresenter.SignInPinUI {
    @Inject
    lateinit var presenter: SignInPinPresenter

    private var pinView: PinView? = null

    private var cleanStart: Int = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        (activity?.application as TransferXApplication).appComponent.inject(this)
        super.onCreate(savedInstanceState)

        arguments?.let { cleanStart = arguments!!.getInt(CLEAN_START) }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.signin_pin_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        pinView = loginPinView
        pinView?.setAnimationEnable(true)

        // Provide reference to base class action button
        buttonNext = signInPinButton
        signInPinButton.isEnabled = false
        signInPinButton.setOnClickListener {
            presenter.checkPinAndLogIn(pinView!!.text?.toString()!!)
            hideKeyboard(pinView)
        }
        textForgotPin.setOnClickListener { presenter.resetPassword() }
    }

    override fun onResume() {
        super.onResume()
        presenter.attachUI(this)
        pinView?.addTextChangedListener(textWatcher)
    }

    override fun onPause() {
        super.onPause()
        presenter.detachUI()

        pinView?.removeTextChangedListener(textWatcher)
    }

    override fun goToPreviousScreen() {
        if (cleanStart == 1) startActivity(context?.intentFor<MainActivity>()?.newTask()?.clearTask())
        activity?.finish()

    }

    override fun goToWelcomeScreen() {
        // TBDD
        Toast.makeText(activity, "This action has to be discussed and debated", Toast.LENGTH_LONG).show()
    }

    override fun showErrorPin() {
        pinView?.setText(Constants.EMPTY)
        setLinesColor(R.color.red)
        pinView?.requestFocus()

        val snackbar = Snackbar.make(view!!, getString(R.string.no_match_pin), Snackbar.LENGTH_LONG)
        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }


    override fun showError(message: String) {
        val snackbar = Snackbar.make(view!!, getString(R.string.generic_error), Snackbar.LENGTH_LONG)
        snackbar.view.setBackgroundColor(Color.RED)
        snackbar.show()
    }

    override fun goToWelcome() {
        //no op
    }

    override fun tagName(): String {
        return SignInPinFragment::class.java.simpleName
    }

    override fun setButtonEnabled(isEnabled: Boolean) {
        setButtonStatus(isEnabled)
    }

    override fun setButtonStatus(isEnabled: Boolean) {
        buttonNext.isEnabled = isEnabled
        buttonNext.background = if (isEnabled) ContextCompat.getDrawable(context!!, R.drawable.btn_style_bright)
        else ContextCompat.getDrawable(context!!, R.drawable.btn_disabled)
    }


    private fun resetErrorPin() {
        setLinesColor(R.color.black)
    }

    private fun setLinesColor(@ColorRes color: Int) {
        pinView?.setLineColor(ContextCompat.getColor(pinView!!.context, color))
    }

    private val textWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            resetErrorPin()
            s.let { presenter.setPinValue(s.toString()) }
        }

        override fun afterTextChanged(s: Editable) {

        }
    }
}
